src/main/java/com/thealgorithms/datastructures/heaps/LeonardoHeap.java

```diff
@@ -0,0 +1,173 @@
+            levelTracker = SingleBitOperations.clearBit(levelTracker, consecutiveTreeIndices.get(0));
+            levelTracker = SingleBitOperations.clearBit(levelTracker, consecutiveTreeIndices.get(1));
+            levelTracker = SingleBitOperations.setBit(levelTracker, consecutiveTreeIndices.get(1) + 1);
+        } else if ((levelTracker & 2) == 0) {
+            levelTracker = SingleBitOperations.setBit(levelTracker, 1);
+        } else {
+            levelTracker = SingleBitOperations.setBit(levelTracker, 0);
+        }
+    }
+
+    private void maxHeapifyTree(int rootNodeIndex, int currentLevel) {
+        // A leonardo tree of level n is just 1 node(the root) plus the leonardo tree of n-1 level(left child) plus leonardo tree of n-2 level(right child)
+        // To maxheapify a leonardo tree we need to compare the current root and roots of it's left and right subtree
+        // We recursively hepify the left and right subtrees using the currentLevel
+
+        // BASE CASE
+        if (currentLevel == 0 || currentLevel == 1) {
+            return; // Trees with one node are in already max-heapified.
+        }
+
+        int currentRootNodeIndex = rootNodeIndex;
+        int rightChildIndex = rootNodeIndex - 1;
+        int leftChildIndex = rootNodeIndex - LeonardoNumber.leonardoNumber(currentLevel - 2) - 1;
+        int childIndexForSwap = -1;
+
+        if (heap.get(rightChildIndex).compareTo(heap.get(leftChildIndex)) >= 0) {
+            childIndexForSwap = rightChildIndex;
+        } else {
+            childIndexForSwap = leftChildIndex;
+        }
+
+        if (heap.get(childIndexForSwap).compareTo(heap.get(currentRootNodeIndex)) > 0) {
+            swap(currentRootNodeIndex, childIndexForSwap);
+            if (childIndexForSwap == rightChildIndex) {
+                maxHeapifyTree(rightChildIndex, currentLevel - 2);
+            } else { // swap happened with the left child
+                maxHeapifyTree(leftChildIndex, currentLevel - 1);
+            }
+        }
+    }
+
+    private void shiftRootAndRestoreHeap() {
+
+        if (heap.isEmpty()) {
+            return;
+        }
+
+        Integer[] currentTreeLevels = LeonardoHeapHelper.findAllTreeIndices(levelTracker);
+        int previousTreeSizeCumulative = 0;
+        ArrayList<Integer> rootNodeIndices = new ArrayList<Integer>();
+
+        // The number of roots are going to be same the the number of levels
+        // iterate over the currentTreeLevels and get roots
+
+        for (int i = 0; i < currentTreeLevels.length; i++) {
+            rootNodeIndices.add(previousTreeSizeCumulative + LeonardoNumber.leonardoNumber(currentTreeLevels[i]) - 1);
+            previousTreeSizeCumulative = previousTreeSizeCumulative + LeonardoNumber.leonardoNumber(currentTreeLevels[i]);
+        }
+
+        int rootNodeIndexForHeapify = rootNodeIndices.getLast();
+        int treeLevelforHeapify = currentTreeLevels[currentTreeLevels.length - 1];
+        boolean swaped = false;
```

Comment:
```suggestion
        boolean swapped = false;
```
You can consider moving this declaration _into_ the loop (smaller scope is always better). This should work, because this variable is always `false` at the end of of each loop.